### PR TITLE
Gitlab runner inventory plugin

### PR DESCRIPTION
**Pull request by morph027**
##### SUMMARY
This adds a dynamic inventory for Gitlab CI runners.

##### ISSUE TYPE

 - Feature Pull Request

##### ANSIBLE VERSION

```
2.8.0
```

### CODE REVIEW

File: lib/ansible/plugins/inventory/gitlab_runners.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    HAS_GITLAB = False
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable):
+    ''' Host inventory parser for ansible using Gitlab API as source. '''
+
+    NAME = 'gitlab_runners'
+
+    def _populate(self):
+        self.gl = gitlab.Gitlab(self.get_option('server_url'), private_token=self.get_option('api_token'))
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/gitlab_runners.py]: I don't know the gitlab API, but shouldn't you capture exceptions at this point?

File: lib/ansible/plugins/inventory/gitlab_runners.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                # Use constructed if applicable
+                strict = self.get_option('strict')
+                # Composed variables
+                self._set_composite_vars(self.get_option('compose'), self.host_attrs, self.host, strict=strict)
+                # Complex groups based on jinja2 conditionals, hosts that meet the conditional are added to group
+                self._add_host_to_composed_groups(self.get_option('groups'), self.host_attrs, self.host, strict=strict)
+                # Create groups based on variable values and add the corresponding hosts to it
+                self._add_host_to_keyed_groups(self.get_option('keyed_groups'), self.host_attrs, self.host, strict=strict)
+        except Exception as e:
+            raise AnsibleError('Unable to fetch hosts from Gitlab API, this was the original exception: %s' % to_native(e))
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/gitlab_runners.py]: return AnsibleParserError so other inventory sources can continue

File: lib/ansible/plugins/inventory/gitlab_runners.py (hunk 1 above)
Comment (morph027) [reply lib/ansible/plugins/inventory/gitlab_runners.py]: :see_no_evil: Sure and absolutely right.

Comment (morph027) [reply lib/ansible/plugins/inventory/gitlab_runners.py]: Ah, no initiating the object does not fail if invalid, it's the first request to the api like `gl.runners.all()` which fails.

**Comment by morph027** [comment]
Adding metadata as well:

```json
{
    "_meta": {
        "hostvars": {
            "30": {
                "ansible_host": "127.0.0.1",
                "gitlab_runner_attributes": {
                    "access_level": "not_protected",
                    "active": true,
                    "architecture": "amd64",
                    "contacted_at": "2019-03-16T09:50:07.779+01:00",
                    "description": "foo_shell",
                    "groups": [],
                    "id": 30,
                    "ip_address": "172.17.0.2",
                    "is_shared": true,
                    "locked": true,
                    "maximum_timeout": null,
                    "name": "gitlab-runner",
                    "online": false,
                    "platform": "linux",
                    "projects": [],
                    "revision": "4745a6f3",
                    "run_untagged": false,
                    "status": "offline",
                    "tag_list": [
                        "shell"
                    ],
                    "token": "aLC3u_ProH3GP67jqNhy",
                    "version": "11.8.0"
                }
            },
            "31": {
                "ansible_host": "127.0.0.1",
                "gitlab_runner_attributes": {
                    "access_level": "not_protected",
                    "active": true,
                    "architecture": "amd64",
                    "contacted_at": "2019-03-16T09:50:42.221+01:00",
                    "description": "foo_docker",
                    "groups": [],
                    "id": 31,
                    "ip_address": "172.17.0.3",
                    "is_shared": true,
                    "locked": false,
                    "maximum_timeout": null,
                    "name": "gitlab-runner",
                    "online": false,
                    "platform": "linux",
                    "projects": [],
                    "revision": "4745a6f3",
                    "run_untagged": false,
                    "status": "offline",
                    "tag_list": [
                        "docker"
                    ],
                    "token": "6UpGTm_AGW_kkaouX96g",
                    "version": "11.8.0"
                }
            },
            "32": {
                "ansible_host": "127.0.0.1",
                "gitlab_runner_attributes": {
                    "access_level": "not_protected",
                    "active": true,
                    "architecture": "amd64",
                    "contacted_at": "2019-03-16T16:58:07.610+01:00",
                    "description": "18c24e5a5ca5",
                    "groups": [],
                    "id": 32,
                    "ip_address": "172.17.0.4",
                    "is_shared": true,
                    "locked": true,
                    "maximum_timeout": null,
                    "name": "gitlab-runner",
                    "online": false,
                    "platform": "linux",
                    "projects": [],
                    "revision": "4745a6f3",
                    "run_untagged": false,
                    "status": "offline",
                    "tag_list": [
                        "docker"
                    ],
                    "token": "hoSoa48t4fQHyuJyVeP3",
                    "version": "11.8.0"
                }
            }
        }
    },
    "all": {
        "children": [
            "description_18c24e5a5ca5",
            "description_foo_docker",
            "description_foo_shell",
            "ungrouped"
        ]
    },
    "description_18c24e5a5ca5": {
        "hosts": [
            "32"
        ]
    },
    "description_foo_docker": {
        "hosts": [
            "31"
        ]
    },
    "description_foo_shell": {
        "hosts": [
            "30"
        ]
    }
}
```

File: lib/ansible/plugins/inventory/gitlab_runners.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
```
Comment (Shaps) [inline lib/ansible/plugins/inventory/gitlab_runners.py]: @morph027, I still see `AnsibleError`?

**Comment by Shaps** [comment]
@morph027 Thanks for this. 

shipit

**Comment by Shaps** [comment]
@Lunik do you have anything against this?

File: lib/ansible/plugins/inventory/gitlab_runners.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable):
+    ''' Host inventory parser for ansible using Gitlab API as source. '''
+
+    NAME = 'gitlab_runners'
+
+    def _populate(self):
+        self.gl = gitlab.Gitlab(self.get_option('server_url'), private_token=self.get_option('api_token'))
+        self.inventory.add_group('all')
```
Comment (resmo) [inline lib/ansible/plugins/inventory/gitlab_runners.py]: not a blocker, but I don't think this is required, it is a default group, instead I would suggest to add a group `self.inventory.add_group('gitlab_runners')` and in the host loop, put them into this group `self.inventory.add_host(host=self.host, group='gitlab_runners')`

**Review by resmo — APPROVED** [review]
Minor note, but not a blocker, I recently found out that a top group named like the plugin may be helpful.
 
LGTM

File: lib/ansible/plugins/inventory/gitlab_runners.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            else:
+                self.runners = self.gl.runners.all()
+            for self.runner in self.runners:
+                self.host = str(self.runner['id'])
+                self.ip_address = self.runner['ip_address']
+                self.host_attrs = vars(self.gl.runners.get(self.runner['id']))['_attrs']
+                self.inventory.add_host(self.host)
+                self.inventory.set_variable(self.host, 'ansible_host', self.ip_address)
+                if self.get_option('verbose_output', True):
+                    self.inventory.set_variable(self.host, 'gitlab_runner_attributes', self.host_attrs)
```
Comment (Lunik) [inline lib/ansible/plugins/inventory/gitlab_runners.py]: If found this part weird with using `self` for temporary variables

Comment (morph027) [reply lib/ansible/plugins/inventory/gitlab_runners.py]: You mean `self.host`? Should i change it?

**Comment by resmo** [comment]
Thanks!

**Pull request merged**